src/main/java/me/chrr/scribble/mixin/BookEditScreenMixin.java

```diff
@@ -497,31 +597,91 @@ private void setPageContent(String newContent) {
         Scribble.LOGGER.warn("setPageContent() was called, but ignored.");
     }
 
+    @Inject(method = "charTyped", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/util/SelectionManager;insert(Ljava/lang/String;)V"), cancellable = true)
+    private void charTypedEditMode(char chr, int modifiers, CallbackInfoReturnable<Boolean> cir) {
+        Command command = new InsertTextCommand<>(this, currentPageSelectionManager, Character.toString(chr));
+        commandManager.execute(command);
+        cir.setReturnValue(true);
+        cir.cancel();
+    }
+
     @Inject(method = "keyPressedEditMode", at = @At(value = "HEAD"), cancellable = true)
     private void keyPressedEditMode(int keyCode, int scanCode, int modifiers, CallbackInfoReturnable<Boolean> cir) {
-        // Copy/cut/paste without formatting when SHIFT is held down.
-        if (hasControlDown() && hasShiftDown() && !hasAltDown()) {
-            if (keyCode == GLFW.GLFW_KEY_C) {
-                this.getRichSelectionManager().copyWithoutFormatting();
-            } else if (keyCode == GLFW.GLFW_KEY_X) {
-                this.getRichSelectionManager().cutWithoutFormatting();
-            } else if (keyCode == GLFW.GLFW_KEY_V) {
-                this.getRichSelectionManager().pasteWithoutFormatting();
+        // Override copy-without-formatting shortcut
+        if (hasControlDown() && hasShiftDown() && !hasAltDown() && keyCode == GLFW.GLFW_KEY_C) {
+            getRichSelectionManager().copyWithoutFormatting();
+            cir.setReturnValue(true);
+            cir.cancel();
+            return;
+        }
+
+        // Override CUT and CUT-without-formatting shortcut
+        if (hasControlDown() && !hasAltDown() && keyCode == GLFW.GLFW_KEY_X) {
+            Command command = new ActionCommand<>(this, () -> {
+                if (hasShiftDown()) {
+                    getRichSelectionManager().cutWithoutFormatting();
+                } else {
+                    getRichSelectionManager().cut();
+                }
+            });
+            commandManager.execute(command);
+            cir.setReturnValue(true);
+            cir.cancel();
+            return;
+        }
+
+        // Override PASTE and PASTE-without-formatting shortcut
+        if (hasControlDown() && !hasAltDown() && keyCode == GLFW.GLFW_KEY_V) {
+            // ToDo try to find more elegant way how to implement PASTE command
+            //  which won't use internal selectionManager.paste/pasteWithoutFormatting() implementation
+            String textToPaste = hasShiftDown() ? Formatting.strip(getRawClipboard()) : getRawClipboard();
+            Command command = new InsertTextCommand<>(this, getRichSelectionManager(), textToPaste);
```

Comment:
Same as above, it's not elegant, it works for now but it would be better if we could somehow circumvent the command manager for certain commands.

Reply:
If we want to track the history of user actions - we need to use Command Manager. I'm not sure what do you mean by "circumvent the command manager" and how it could help here?

Also could you add more details why cut/paste undo/redo is not correct? some example would be helpfull.
> As for pasting (and cutting), I feel there's a need for some way an action is different when you redo it vs when it first happens. Especially in the event of cutting, where the first time it should cut, but the second time it should just insert text.

Reply:
What I mean is that undoing and redoing shouldn't interact with the clipboard at all. Redoing a cut shouldn't put it back on your clipboard, it doesn't do that in any other program. Same for pasting, redoing a paste shouldn't paste from the clipboard, but just insert the text it originally pasted (as it does right now, where you worked around selectionManager.paste).

A more elegant way would be to have some way to execute the normal cut/paste when the action is first done, but just deleting / inserting the right text when the command is called on later with an undo/redo. I don't know that specifically would be perfect, but it would prevent these problems.

Reply:
Thanks for the explanation. I [decided to use](https://github.com/chrrs/scribble/pull/18/commits/0d886b62a07adeebb0d4e6217128683e97fc3740) the same approach I used for the Paste command. Exposing "if the command executes for the first time" to the `Command::execute` would break the abstraction for Command interface. I would like to avoid it.

I still don't understand why Mojang decided to let SelectionManager interact with the clipboard. It should be outside of it, in my opinion😬